### PR TITLE
Add role validation to basicauth checking

### PR DESCRIPTION
**Pull request by slizco**
## Reasons for making these changes

We'd like to extend the `basicauth` capabilities to include validation of role-based access. This [ADR](https://github.com/heroku/runtime/pull/4837) gives context.

## Description of changes

- Adds a `RoleValidator` interface that provides a `Validate(role, method string) bool` func that should return true when the `role` has access to `method`.
- Modifies `basicauth.GRPCAuthFunc` to call `roleValidator.Validate` if role validator is defined.
- Adds testing of the above.

## Runtime issue
https://trello.com/c/eEdSQMMa/193-implement-role-validator-in-heroku-x-basicauth

### CODE REVIEW

**Comment by danp** [comment]
This maybe trends from authentication to authorization, and I'd say so far basicauth has only been about authentication. Should authorization happen in a more distinct place to keep it separate from authentication?

**Comment by slizco** [comment]
@danp Yes, you're right, it is more authorization. We'd like to use the grpc interceptor model to accomplish this though, and this is the simplest way for us to do so. We could write a a whole separate interceptor if need be though. I do think adding the role validation as an optional functionality here is okay. I'll let others weigh in too

File: hmiddleware/basicauth/checker.go
Comment (galbacarys) [inline hmiddleware/basicauth/checker.go:54]: Is it possible we could want more than one role validator?

Comment (slizco) [reply hmiddleware/basicauth/checker.go:54]: Mm I can't see a reason we would. The Credentials field is a list because they are pairs of username/passwords, where each client gets its own pair.

**Review by galbacarys — APPROVED** [review]
LGTM pending ci passing. I think the conflation of AuthN/AuthZ isn't the biggest deal in the world since it's such a simple scheme and the authZ portion is optional.

**Comment by slizco** [comment]
Yeah the failure is due to test coverage I believe. I'm going to add a couple more cases

File: hmiddleware/basicauth/grpc.go
Comment (ypaq) [inline hmiddleware/basicauth/grpc.go:43]: Is the error message relayed back to the client?

**Review by ypaq — APPROVED** [review]
Looks good to me, thank you.

**Comment by ypaq** [comment]
To @danp's point - I agree that they are different concerns that we might want to consider to separate. Alternatively to the implementation here, the grpc_middleware supports chaining of interceptors which we could utilize to separate the concerns of authn and authz. We could leverage the context to pass the role. If that would be a better implementation I'm not sure. The fact that we're using basic auth is a tirefire on its own, and this change definitely doesn't make it worse.

**Review by borncrusader — APPROVED** [review]
LGTM. Great work, @slizco

File: hmiddleware/basicauth/grpc.go
Comment (iamjem) [inline hmiddleware/basicauth/grpc.go:49]: you can use `%q` instead of escaping the quotes here

**Review by iamjem — COMMENTED** [review]
I have similar feels as @danp, its common for authentication and authorization to exist in separate layers/phases of request processing (the order matters of course!). So the authz part would just be a separate unary/stream interceptor, and then the checker type could be left as-is. There's already some utilities for collecting grpc options--including interceptors--and we could just have this be another optional interceptor to be provided and included in the chain.

I also don't think we need to put everything that feels general purpose in heroku/x immediately, we can keep it local to a project while we iterate and polish the idea, and move it here when it makes sense (for example when multiple projects need access to said functionality).

**Comment by slizco** [comment]
We do have this same code in `heroku/runtime/lib`, so I'm okay with expanding upon that for now. I'm going to work on writing this as a separate interceptor over there and see how that looks

File: hmiddleware/basicauth/grpc.go
Comment (slizco) [reply hmiddleware/basicauth/grpc.go:43]: Mmm good question. Don't know, but I can check what we do

**Comment by slizco** [comment]
Closing this stale PR. The role-based authorization functionality has been added to `heroku/runtime/lib` instead.

**Pull request closed without merging**
